### ISSUE TITLE
[BUG] Open non-UTF8 file on macOS causes RStudio broken

### ISSUE THREAD

**Issue by shrektan** [rstudio/rstudio#6709]
### System details

    RStudio Edition : Desktop
    RStudio Version : Version 1.4.200
    OS Version      : macOS Catalina 10.15.3
    R Version       :  3.6.2

### Steps to reproduce the problem

Just run the below code.

```r
tbl <- data.frame(
  CN = c('国家电网公司', '中国联合网络通信股份有限公司', '中华人民共和国铁道部', '中国南方电网有限责任公司'),
  stringsAsFactors = FALSE
)
write.csv(tbl, file = "./gb2312.csv", fileEncoding = 'GB2312', row.names = FALSE)
file.edit('./gb2312.csv')
```

### Describe the problem in detail

On macOS, the default encoding is UTF-8. If the user accidentally opens a text file that's not encoded in UTF-8, e.g., GB2312, the RStudio IDE will crash. And you can never open that project any longer unless the hidden `.Rproj.user` folder gets removed.

(OK, I know I should provide a diagnose report but I believe it's very easy to reproduce. I'm happy to provide if any problem)

### Describe the behavior you expected

No error.

### Screenshot of the error

<img width="1424" alt="image" src="https://user-images.githubusercontent.com/8368933/79997328-0a68a580-84ec-11ea-8466-1049d18ca757.png">

<img width="444" alt="image" src="https://user-images.githubusercontent.com/8368933/79997345-105e8680-84ec-11ea-8759-c4c293a50b77.png">

<img width="327" alt="image" src="https://user-images.githubusercontent.com/8368933/79997368-15233a80-84ec-11ea-9462-d36eb65e9892.png">

**Comment by kevinushey** [comment]
Thanks for the bug report -- I've reproduced this as well and will take a look.

**Comment by kevinushey** [comment]
Backtrace:

```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 3.2
  * frame #0: 0x000000010444d066 rsession`rstudio::session::Options::read(int, char* const*, std::__1::basic_ostream<char, std::__1::char_traits<char> >&) [inlined] rstudio::core::ProgramStatus::exit(this=<unavailable>) const at ProgramStatus.hpp:47:31 [opt]
    frame #1: 0x000000010444d066 rsession`rstudio::session::Options::read(this=0x0000000104e77408, argc=9, argv=0x00007ffeeb98b9a8, osWarnings=0x00007ffeeb98b240) at SessionOptions.cpp:505 [opt]
    frame #2: 0x00000001043971fb rsession`main(argc=9, argv=0x00007ffeeb98b9a8) at SessionMain.cpp:1738:38 [opt]
    frame #3: 0x00007fff6f28fcc9 libdyld.dylib`start + 1
    frame #4: 0x00007fff6f28fcc9 libdyld.dylib`start + 1
```

**Comment by kevinushey** [comment]
Here's a brief summary of the issue, as I understand it:

1. We attempt to open the document, assuming the encoding is UTF-8. The document is opened, but most of the characters cannot be correctly decoded.

2. Our editor instance replaces the characters that could not be decoded with `?`. The size of these characters is not in sync with the internal (UTF-8) representation of these characters.

3. The IDE attempts to save a diff of the document. This attempt fails, because of the editor's notion of the document is out-of-sync with the server-side notion of the document.

There are a few things that need to happen here, I think:

1. We need to signal when the attempts to read / decode the document fail;
2. We need to more gracefully handle the case where the attempt to save a diff for this document fails;
3. We need to eliminate the spurious document save that happens when the document is first opened.

**Comment by kevinushey** [comment]
Thank you for the bug report and sorry for the trouble!

**Comment by rich-rstudio** [comment]
verified fixed in Version 1.3.947
